source/guides/architecture/interactors.md

```diff
@@ -0,0 +1,391 @@
+---
+title: Architecture - Interactors
+---
+
+# Overview
+Hanami provides an **optional** tool for organizing your code.
+
+These are *Interactors*, also referred to as *service objects* or *operations*
+
+We provide `Hanami::Interactor`,
+but you can use Plain Old Ruby Objects or a different library.
+
+# A new feature: User Sign-Up
+We want to add users to our `bookshelf` application.
+
+Note: this is just to explain `Hanami::Interactor` and
+we will not be building a full authentication system.
+We are _only_ concerning ourselves with signing up,
+not logging in or any other features.
+
+This is a good example of functionality to extract into an Interactor,
+because user sign-up is not trivial.
+
+
+# Background
+Passwords should **never** be stored in plain-text.
+
+We should first _hash_ the password.
+This is a function that always returns the same random output for the
+same input.
+It is _one-way_,
+which means it's extremely hard to reverse it,
+that is, to get the password from the hashed output.
+(This output is called a digest.)
+
+We don't store plain-text passwords because if our database is compromised,
+we don't want to reveal our users' passwords
+(since many people re-use passwords).
+
+Instead, attackers would only get the hashed results.
+
+Note: this approach is still vulnerable to
+[Rainbow tables](https://en.wikipedia.org/wiki/Rainbow_table).
+
+In order to combat this, we'll use a hashing scheme that employs a
+[`salt`](https://en.wikipedia.org/wiki/Salt_(cryptography)).
+
+# Preparing
+Let's say we have our `bookshelf` application,
+from the [Getting Started]((/guides/getting-started)
+and we want to add a feature.
+
+Clone the `bookshelf` application so we're starting from the same place.
+```bash
+git clone git@github.com:hanami/bookshelf.git hanami-bookshelf
+cd hanami-bookshelf
+bundle install
+HANAMI_ENV=development hanami db prepare
+HANAMI_ENV=test hanami db prepare
+bundle exec rake # All the tests should pass
+```
+
+# Adding User model
+```bash
+hanami generate model User
+```
+
+You'll see an entity (`User`) and a repository (`UserRepository`) created,
+as well as their tests files.
+
+You'll also see a migration was generated,
+something like:
+`db/migrations/20170326152126_create_users.rb`
+
+Edit this file:
+
+```ruby
+Hanami::Model.migration do
+  change do
+    create_table :users do
+      primary_key :id
+      string :email, null: false, unique: true
+      string :password_digest
+
+      column :created_at, DateTime, null: false
+      column :updated_at, DateTime, null: false
+    end
+  end
+end
+```
+
+We'll add a couple columns:
+- `email`, we require this field, and make sure it's unique
+- `password_digest`, this will be the user's hashed password
+
+
+Let's create that table:
+
+```bash
+bundle exec hanami db prepare
+```
+
+# Interactor
```

Comment:
@jodosha just a idea:
maybe we introduce new cli command `hanami generate interactor <interactors_name>` ?